miden/src/cli/debug.rs

```diff
@@ -0,0 +1,219 @@
+                    }
+                    // rewind to the beginning of program execution
+                    "!rewind" => {
+                        while let Some(prev_vm_state) = prev_vm_state(&mut vm_state_iter) {
+                            vm_state = prev_vm_state;
+                        }
+                    }
+                    // rewind a specified number of cycles - !rewind.num_cycles
+                    x if x.len() > 8 && &x[..8] == "!rewind." => {
+                        let command = x.to_string();
+                        let parts: Vec<&str> = command.split(".").collect();
+                        let num_cycles = parts[1].parse::<usize>();
+                        match num_cycles {
+                            Ok(cycles) => {
+                                for _cycle in 0..cycles {
+                                    match prev_vm_state(&mut vm_state_iter) {
+                                        Some(next_vm_state) => {
+                                            vm_state = next_vm_state;
+                                        }
+                                        None => break,
+                                    }
+                                }
+                            }
+                            Err(err) => {
+                                println!("Could not parse number of cycles! {err}");
+                            }
+                        }
+                    }
+                    // print memory
+                    "!mem" => {
+                        for (address, mem) in vm_state.memory.iter() {
+                            print_mem_address(*address, &mem)
+                        }
+                    }
+                    // print memory at address - !mem[address]
+                    x if x.len() > 6
+                        && &x[..5] == "!mem["
+                        && x.chars().nth(x.len() - 1) == Some(']') =>
+                    {
+                        let address = x[5..x.len() - 1].parse::<u64>();
+                        match address {
+                            Ok(address) => {
+                                let mem = vm_state.memory.iter().find_map(|(addr, mem)| {
+                                    match address == *addr {
+                                        true => Some(mem),
+                                        false => None,
+                                    }
+                                });
+
+                                match mem {
+                                    Some(mem) => print_mem_address(address, mem),
+                                    None => println!("memory at address '{}' not found", address),
+                                }
+                            }
+                            Err(err) => println!("Could not parse memory address! {err}"),
+                        }
+                    }
+                    // print stack
+                    "!stack" => print_stack(&vm_state.stack),
```

Comment:
another note: it would also be useful to have the stack offset available. The current output is:

```
>> !stack
1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0
```

which is easy for output like the above, but gets a bit unwindly if we have to find the 8th element. Here is a proposed format:

```
[0] = 1
[1] = 1
[2] = 0
...
```

Reply:
It would also be nice to have access to the rest of the stack, having `!stack[<number>]` command could be nice here.

Reply:
done